hare3/algorithm.go

```diff
@@ -0,0 +1,478 @@
+		if p.hardLocked[j] {
+			mm = &OutputMessage{
+				round:  NewAbsRound(j, 5),
+				values: []types.Hash20{*p.Li[j]},
+			}
+		} else {
+			for _, c := range candidates {
+				candidateHash := toHash(c.values)
+				// Check to see if valid proposal for this iteration
+				// Round 5 condition c
+				_, ok := p.Ti[j][candidateHash]
+				if !ok {
+					continue
+				}
+				// Check leader
+				// Round 5 condition d
+				if !p.lc.IsLeader(c.id, NewAbsRound(j, 2)) {
+					continue
+				}
+				// Check grade
+				// Round 5 condition e
+				if c.grade < 2 {
+					continue
+				}
+				// Check subset of threshold values at round 3
+				// Round 5 condition f
+				if !isSubset(c.values, p.Vi[3]) {
+					continue
+				}
+				// Check superset of highest graded values or we received a
+				// commit for the previous iteration from thresh-gossip for
+				// this set with grade >= 1.
+				// Round 5 condition g
+				if !isSubset(p.Vi[5], c.values) {
+					// Check for received message
+					lastIterationCommit := NewAbsRound(j-1, 5)
+					values := p.tgg.RetrieveThresholdMessages(lastIterationCommit, 1)
+
+					// If the candidate is not superset of highest graded
+					// values and no commit for that candidate was received in
+					// the previous iteration with grade >= 1 then we go to the
+					// next candidate.
+					if !slices.Contains(values, candidateHash) {
+						continue
+					}
+				}
+				// Locked value for this iteration is nil or matches set
+				//
+				// Round 5 condition h
+				if !(p.Li[j] == nil || *p.Li[j] == candidateHash) {
+					continue
+				}
+				mm = &OutputMessage{
+					round:  NewAbsRound(j, 5),
+					values: []types.Hash20{candidateHash},
+				}
+				break
+			}
+		}
+		return mm, nil
+	case Notify:
+		// Case 1
+		values := p.tgg.RetrieveThresholdMessages(NewAbsRound(j-1, 6), 5)
+		resultHash, result := matchInPreviousIterations(values, p.Ti, j)
+		if resultHash != nil {
+			return &OutputMessage{
+				round:  NewAbsRound(j, 6),
+				values: []types.Hash20{*resultHash},
+			}, result
+		}
+		// Case 2
+		if active {
+			values := p.tgg.RetrieveThresholdMessages(NewAbsRound(j, 5), 5)
+			setHash, _ := matchInIteration(values, p.Ti, j)
+			return &OutputMessage{
+				round:  NewAbsRound(j, 6),
+				values: []types.Hash20{*setHash},
+			}, nil
+
+		}
+	}
+
+	return nil, nil
+}
+
+func hashBytes(v []byte) types.Hash20 {
+	return types.Hash20{}
+}
+
+type OutputMessage struct {
+	round  AbsRound
+	values []types.Hash20
+}
+
+// with int8 we have a max iterations of 17 before we overflow.
+// AbsRound (short for absolute round) represents a round of the hare protocol. The hare protocol progresses in iterations and there are 7 rounds per iteration, except for the first iteration which has a one off pre-round  the abs round starts at -1 and increments for each successive round.
+//
+// from the abs round both round and message type can be
+// the iteration and the round within that iteration can be derived.
+type AbsRound int8
```

Comment:
this needs to be larger (max number of iterations is 40) and unsigned if it is going to be encoded with scale. note that we use compact encoding with scale, where integer is encoded into 1, 2, 4, 9 bytes depending on its size

Reply:
Thank you, I was wondering if this would be sufficient